core/src/ping_pong.rs

```diff
@@ -0,0 +1,263 @@
+    }
+
+    fn signable_data(&self) -> Cow<[u8]> {
+        Cow::Owned(self.hash.as_ref().into())
+    }
+
+    fn get_signature(&self) -> Signature {
+        self.signature
+    }
+
+    fn set_signature(&mut self, signature: Signature) {
+        self.signature = signature;
+    }
+}
+
+impl PingCache {
+    pub fn new(ttl: Duration, cap: usize) -> Self {
+        Self {
+            ttl,
+            pings: LruCache::new(cap),
+            pongs: LruCache::new(cap),
+            cache: LruCache::new(cap),
+        }
+    }
+
+    /// Checks if the pong hash, pubkey and socket match a ping message sent
+    /// out previously. If so records current timestamp for the remote node and
+    /// returns true.
+    /// Note: Does not verify the signature.
+    pub fn add(&mut self, pong: &Pong, socket: SocketAddr, now: Instant) -> bool {
+        let node = (pong.pubkey(), socket);
+        match self.cache.peek(&pong.hash) {
+            Some(value) if *value == node => {
+                self.pings.pop(&node);
+                self.pongs.put(node, now);
+                self.cache.pop(&pong.hash);
+                true
+            }
+            _ => false,
+        }
+    }
+
+    /// Checks if the remote node has been pinged recently. If not, calls the
+    /// given function to generates a new ping message, records current
+    /// timestamp and hash of ping token, and returns the ping message.
+    fn maybe_ping<T, F>(
+        &mut self,
+        now: Instant,
+        node: (Pubkey, SocketAddr),
+        mut pingf: F,
+    ) -> Option<Ping<T>>
+    where
+        T: Serialize,
+        F: FnMut() -> Option<Ping<T>>,
+    {
+        // Rate limit consecutive pings sent to a remote node.
+        let delay = self.ttl / 64;
+        match self.pings.peek(&node) {
+            Some(t) if now.saturating_duration_since(*t) < delay => None,
+            _ => {
+                let ping = pingf()?;
+                let hash = hash::hash(&serialize(&ping.token).ok()?);
+                self.pings.put(node, now);
+                self.cache.put(hash, node);
+                Some(ping)
+            }
+        }
+    }
+
+    /// Returns true if the remote node has responded to a ping message.
+    /// Removes expired pong messages. In order to extend verifications before
+    /// expiration, if the pong message is not too recent, and the node has not
+    /// been pinged recently, calls the given function to generates a new ping
+    /// message, records current timestamp and hash of ping token, and returns
+    /// the ping message.
+    /// Caller should verify if the socket address is valid. (e.g. by using
+    /// ContactInfo::is_valid_address).
+    pub fn check<T, F>(
+        &mut self,
+        now: Instant,
+        node: (Pubkey, SocketAddr),
+        pingf: F,
+    ) -> (bool, Option<Ping<T>>)
+    where
+        T: Serialize,
+        F: FnMut() -> Option<Ping<T>>,
+    {
+        let (check, ping) = match self.pongs.get(&node) {
```

Comment:
nit: rename ping-> should_ping to communicate it's a boolean

Reply:
done